doc/designs/quic-design/congestion-control.md

```diff
@@ -0,0 +1,229 @@
+Congestion control API design
+=============================
+
+This is mostly inspired by the MSQUIC congestion control API
+as it was the most isolated one among the libraries.
+
+The API is designed in a way to be later easily transformed into a
+a fetchable implementation API.
+
+The API is centered around two structures - the `OSSL_CC_METHOD`
+structure holding the API calls into the congestion control module
+and `OSSL_CC_DATA` opaque type that holds the data of needed
+for the operation of the module.
+
+Most of the information when the functions of the API are
+supposed to be called and how the API is designed to operate
+should be clear from the comments for the individual
+functions in [Appendix A](#appendix-a).
+
+Changeables
+-----------
+
+As some parameters that the congestion control algorithm needs might
+be updated during the lifetime of the connection these parameters
+are called changeables.
+
+The CC implementation will save the pointers to these parameters'
+data and will use the current value of the data at the pointer when
+it needs to recalculate the allowance and whether the sending is
+blocked or not.
+
+Some examples of changeables
+
+ - `payload_length` - the maximum length of payload that can be sent
+   in a single UDP packet (not counting UDP and IP headers).
+ - `smoothed_rtt` - the current round trip time of the connection
+   as computed from the acked packets.
+ - `rtt_variance` - the variance of the `smoothed_rtt` value.
+
+Thread handling
+---------------
+
+The `OSSL_CC_DATA` is created with the `new` function per each
+connection. The expectation is that there is only a single thread
+accessing the `ccdata` which should not be a limitation as the calls
+into the module will be done from the packetizer layer of the QUIC
+connection handling and there the eventual writes from multiple threads
+handling individual streams of the connection have to be synchronized
+already to create the packets.
+
+Congestion event handling
+-------------------------
+
+The congestion state of a connection can change only after some event
+happens - i.e., a packet is considered lost meaning the `on_data_lost()`
+is called, or an ack arrives for a packet causing calls to
+`on_data_acked()` or `on_spurious_congestion_event()` functions.
+The congestion control does not produce any timer events by itself.
+
+Exemptions
+----------
+To facilitate probing and to avoid having to always special-case
+probing packets when considering congestion on sending, the
+`set_exemption()` function allows setting number of packets that
+are allowed to be sent overriding the eventual congestion state.
```

Comment:
> If the intent is the former, what prevents this additional capacity from being used by non-probing packets?

Well, nothing. Just correctness of the calling implementation.

Reply:
That feels like a fragile design to me.  It seems like it would be more robust to shift the burden onto the probe packets to mark themselves as special and thus eligible for an exception than to just blanket allow any extra traffic.

Reply:
At least in MsQuic, is is not so easy because of the levels of abstraction. The logic for deciding that something needs to be sent is separate from actually building and sending the logic.